common/v2/services/Store/DataManager/DataProvider.tsx

```diff
@@ -0,0 +1,104 @@
+import React, { createContext, Dispatch, useReducer, useCallback } from 'react';
+
+import { DataStore, LSKeys } from 'v2/types';
+import { useThrottleFn, useLocalStorage, useEvent } from 'v2/vendor';
+import { addDevSeedToSchema, removeSeedDataFromSchema, getCurrentDBConfig } from 'v2/database';
+import { appDataReducer, ActionV, ActionT, ActionPayload } from './reducer';
+import { ActionFactory } from './actions';
+import { deMarshallState, marshallState } from './utils';
+import { useDatabase } from './useDatabase';
+
+export interface DataCacheManager extends DataStore {
+  createActions(k: LSKeys): ReturnType<typeof ActionFactory>;
+  resetAppDb(): void;
+  addSeedData(): void;
+  removeSeedData(): void;
+}
+
+interface EncryptedStorage {
+  encryptedDb: any;
+  setEncryptedCache(ls: string): void;
+  destroyEncryptedCache(): void;
+  setUnlockPassword(pwd: string): void;
+  getUnlockPassword(): string;
+}
+export type IDataContext = DataCacheManager & EncryptedStorage;
+
+export const DataContext = createContext({} as IDataContext);
+
+export const DataProvider: React.FC = ({ children }) => {
+  /*
+   *  Create the our master store, sync with persistance layer,
+   *  Provide lj
+   */
+  const { db, setDb, resetDb, defaultValues } = useDatabase();
+
+  const [appState, dispatch]: [DataStore, Dispatch<ActionV>] = useReducer(
+    appDataReducer,
+    db, // Initial state
+    marshallState // method to run on initial state
+  );
+
+  const resetAppDb = useCallback(
+    (newDb = defaultValues) => {
+      resetDb(newDb); // Reset the persistence layer
+      dispatch({
+        type: ActionT.RESET,
+        payload: { data: marshallState(newDb) } as ActionPayload<DataStore>
+      }); // Reset the Context
+    },
+    [defaultValues]
+  );
+
+  /*
+   * Manage sync between appState an db
+   */
+  const syncDb = useCallback(() => {
+    console.debug('Updating LocalStorage');
+    setDb(deMarshallState(appState));
+  }, [appState]);
+  // By default we sync no more than once a minute
+  useThrottleFn(syncDb, 60000, [appState]);
+  // In any case we sync before the tab is closed
+  useEvent('beforeunload', syncDb);
```

Comment:
Is this the only trigger that can possibly happen? Does this trigger when all of the following?:
1) Closes the tab
2) Closes the window
3) Kills the browser process using cmd
4) Hard shuts down the computer

Reply:
So according to the Page Lifecycle best practices, I learn that using `beforeunload` is a bad practice: https://developers.google.com/web/updates/2018/07/page-lifecycle-api

Updated the code for `visibilitychange` event and a check for `document.hidden`
https://developer.mozilla.org/en-US/docs/Web/API/Document/visibilityState

This will execute the JS when "document is either a background tab or part of a minimised window, or the OS screen lock is active."

For the other possibilities, I would prefer to address this PR.
Since we're syncing every minute, and that most operations can be either recovered or reproduced by the user I'm confortable with the possibility of missing a sync atm.
If you think the user could perform an important operation within the 60sec, we can maybe reduce the sync time?